README
This project has migrated to https://git.merproject.org/mer-core/pulseaudio-modules-nemo
as part of the Mer/Nemo merge work


debian/changelog
pulseaudio-meego (0.9.19.0-35) unstable; urgency=low

  * voice: Convert byte amounts between different sample spec domains
    correctly.
  * Fixes: NB#297413 - The frame alignment problem is now fixed.

 -- Tanu Kaskinen <tanu.kaskinen@digia.com>  Tue, 21 Feb 2012 17:49:00 +0200

pulseaudio-meego (0.9.19.0-34) unstable; urgency=low

  * cmtspeech: Increase cmtspeech cleanup timer timeout to 5s.
  * Fixes: NB#282955 - Allow more time for network to release audio.

 -- Maemo Multimedia <multimedia@maemo.org>  Thu, 03 Nov 2011 14:15:21 +0200

pulseaudio-meego (0.9.19.0-33) unstable; urgency=low

  * cmtspeech: Adapt to the changed pa_thread_new() interface.
  * Fixes: NB#286317 - Adapt to the pulseaudio changes.
  * Fixes: NB#278727 - Added missing pulseaudio-dev dependency to
    pulseaudio-module-meego-common-dev package.

 -- Maemo Multimedia <multimedia@maemo.org>  Mon, 31 Oct 2011 14:01:44 +0200

pulseaudio-meego (0.9.19.0-32) unstable; urgency=low

  * Fixes: NB#275839 - Fixed assertion failure when voice module master sink
    input moves to sink.null.

 -- Maemo Multimedia <multimedia@maemo.org>  Tue, 04 Oct 2011 13:28:50 +0300

pulseaudio-meego (0.9.19.0-31) unstable; urgency=low

  * Fixes: NB#280746 - cmtspeech connection resetting improved.

 -- Maemo Multimedia <multimedia@maemo.org>  Sat, 10 Sep 2011 12:46:08 +0300

pulseaudio-meego (0.9.19.0-30) unstable; urgency=low

  * Reverted the previous version of cmtspeech forced cleanup patch, and
    replaced it with the original version that is verified to work.
  * Fixes: NB#265431 - Return to normal mode after call even if the cmtspeech
    server "forgets" to give the signal that call has ended.

 -- Maemo Multimedia <multimedia@maemo.org>  Mon, 29 Aug 2011 17:25:12 +0300

pulseaudio-meego (0.9.19.0-29) unstable; urgency=low

  * cmtspeech: Reset cmtspeech connection if cmtspeech server is inactive.
  * Fixes: NB#265431 - Return to normal mode after call even if the cmtspeech
    server "forgets" to give the signal that call has ended.

 -- Maemo Multimedia <mmf-audio@projects.maemo.org>  Wed, 10 Aug 2011 12:54:42 +0300

pulseaudio-meego (0.9.19.0-28) unstable; urgency=low

  * Fixes: NB#255575 - Pulseaudio crash in verbose mode is fixed when Ctrl+C pressed.
  * Fixes: NB#269844 - Recorded audio from SCO is now upmixed properly.
  * Fixes: NB#262783 - Fixed virtual sink rewinding.
  * Fixes: NB#269083 - Fixed memory leaks in Pulseaudio.
  * voice: Reduce debug log level.
  * voice: Upmix stream for proper formating in 8KHz path.
  * voice and music: Sink rewind request is forwarded to master only when the request is larger than 0 bytes.
  * record: Unlink source output and then set thread_info.attached flag to FALSE.
  * common: Deinitialize parameter stop requests.

 -- Maemo Multimedia <mmf-audio@projects.maemo.org>  Wed, 03 Aug 2011 09:27:46 +0300

pulseaudio-meego (0.9.19.0-27) unstable; urgency=low

  * voice: Reset the ear ref loop also when unsuspending a held VoIP call.
  * Fixes: NB#266609 - Echo cancellation keeps working after the remote has continued
    a held call.

 -- Maemo Multimedia <mmf-audio@projects.maemo.org>  Thu, 16 Jun 2011 22:55:52 +0300

pulseaudio-meego (0.9.19.0-26) unstable; urgency=low

  * voice: Handle cases when downlink side was being called multiple
    times in a row.
  * voice: Take render_memblockq size into latency calculations on
    downlink side.
  * voice: Remove the earlier hack for not reseting the ear-ref loop when
    there is no enough data in ear-ref loop.
  * voice: Handle cases when alsa-sink was requesting non hw fragment size
    amount of data.
  * voice: Reset the ear-ref loop when hw_source_output_attach_cb is
    being called.
  * Fixes: NB#254278 - Complete fix for acoustic echo cancellation problems.

 -- Maemo Multimedia <multimedia@maemo.org>  Wed, 15 Jun 2011 10:09:17 +0300

pulseaudio-meego (0.9.19.0-25) unstable; urgency=low

  * voice: Improve stability of AEC reference loop after rewinds.
  * Fixes: NB#254278 - Problems with occasionally non-working acoustic echo
    cancellation (AEC) algorithm during voice call especially in IHF mode
    diminished further.

 -- Maemo Multimedia <multimedia@maemo.org>  Wed, 01 Jun 2011 16:02:18 +0300

pulseaudio-meego (0.9.19.0-24) unstable; urgency=low

  * voice: Reset the ear ref loop when there is a rewind happening on voip
    sink and do not fire the HOOK_AEP_UPLINK hook if there were not enough
    data in the ear ref loop.
  * Fixes: NB#254278 - Problems with occasionally non-working acoustic echo
    cancellation (AEC) algorithm during voice call significantly diminished.

 -- Maemo Multimedia <multimedia@maemo.org>  Tue, 31 May 2011 14:26:42 +0300

pulseaudio-meego (0.9.19.0-23) unstable; urgency=low

  * voice and music: Make sure algotihms are applied to writable memchunk.
  * Fixes: NB#251721 - Samples played from sample-cache do not get corrupted
    anymore.

 -- Maemo Multimedia <multimedia@maemo.org>  Tue, 17 May 2011 15:14:01 +0300

pulseaudio-meego (0.9.19.0-22) unstable; urgency=low

  * algorithm-hook: Don't ref the core.
  * cmtspeech: Check return value of cmtspeech_buffer_to_memchunk().
  * Fixes: NB#253903 - Prevented segmentation fault in some corner cases
    after unsuccessful cellular call.

 -- Maemo Multimedia <multimedia@maemo.org>  Mon, 09 May 2011 15:48:24 +0300

pulseaudio-meego (0.9.19.0-21) unstable; urgency=low

  * Fixes: NB#244880 - Prevented modem crash after simless emergency call.
  * common: Rewrite algorithm-hook interface to better suite algorithm data
    processing.
  * common: Add algorithm-base, a skeleton algorithm data processing module.
  * common-dev: Generate pkg-config file.

 -- Maemo Multimedia <multimedia@maemo.org>  Tue, 26 Apr 2011 15:24:08 +0300

pulseaudio-meego (0.9.19.0-20) unstable; urgency=low

  * sidetone: return when current master volume is same as previous.
  * sidetone: when master sink volume is mute then better to mute the sidetone also

src/common/src-48-to-8.c
/*
 * Copyright (C) 2010 Nokia Corporation.
 *
 * Contact: Maemo MMF Audio <mmf-audio@projects.maemo.org>
 *          or Jyri Sarha <jyri.sarha@nokia.com>
 *
 * These PulseAudio Modules are free software; you can redistribute
 * it and/or modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation
 * version 2.1 of the License.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
 * USA.
 */

#include <string.h>
#include <stdlib.h>
#include "src-48-to-8.h"

#ifdef ARM_DSP
#include <dspfns.h>
#endif

#define DOWNSAMPLING_FACTOR_1 3
#define FILTER_LENGTH_1 16
#define FILTER_MEMORY_1 15
/* FILTER_MEMORY_HOP_1 = (int) FILTER_LENGTH_1 / DOWNSAMPLING_FACTOR_1 */
#define FILTER_MEMORY_HOP_1 5
/* STEREO_DOWNSAMPLING_FACTOR_1 = DOWNSAMPLING_FACTOR_1 * 2 */
#define STEREO_DOWNSAMPLING_FACTOR_1 6

#define DOWNSAMPLING_FACTOR_2 2
#define FILTER_LENGTH_2 80
#define FILTER_MEMORY_2 79
/* (int) FILTER_LENGTH_2 / DOWNSAMPLING_FACTOR_2 */
#define FILTER_MEMORY_HOP_2 40

static short temp_buffer[SRC_48_TO_8_MAX_INPUT_FRAMES / 3];

struct src_48_to_8
{
  short filter_memory_1[FILTER_LENGTH_1 * 2];
  short filter_memory_2[FILTER_LENGTH_2];
};

static const signed short filter_coeffs_1[] =
{
  8, 69, 185, 94, -502, -1527, -2132, -1143,
  1841, 5719, 8467, 8671, 6566, 3651, 1390, 290
};

static const signed short filter_coeffs_2[] =
{
  2, -6, 4, 4, -5, -6, 7, 9, -10, -12, 13, 17,
  -18, -24, 23, 34, -28, -46, 34, 62, -40, -84,
  44, 111, -46, -146, 43, 189, -31, -242, 6, 304,
  42, -370, -126, 421, 261, -376, -313, 366, 333,
  -477, -555, 377, 631, -371, -808, 261, 939, -146,
  -1094, -42, 1218, 277, -1319, -584, 1361, 958,
  -1319, -1403, 1144, 1901, -776, -2414, 131, 2851,
  890, -3020, -2377, 2523, 4284, -585, -5955, -4086,
  4336, 11559, 12142, 7681, 2961, 564
};

src_48_to_8 *alloc_src_48_to_8(void)
{
  src_48_to_8 *src = (src_48_to_8 *) malloc(sizeof(src_48_to_8));
  memset(src, 0, sizeof(*src));
  return src;
}

void free_src_48_to_8(src_48_to_8 *src)
{
  free(src);
}

#ifdef USE_SATURATION
#ifdef ARM_DSP
static inline int EAP_Clip16(int input)
{
  return __ssat( input, 16 );
}
#else
static inline int EAP_Clip16(int input)
{  input = input < (-32768) ? (-32768) : input;
  input = input > 32767 ?  32767 : input;
  return input;
}
#endif
#endif

int process_src_48_to_8(src_48_to_8 *s,
			short *output,
			short *input,
			int input_frames)
{
    signed short *input_samples = 0;
    signed short *poly_start = 0;
    signed short *output_samples = 0;
    signed int result = 0;
    int intermediate_frames = input_frames / 3;
    int output_frames = input_frames / 6;
    int samples_to_process_1 = intermediate_frames - FILTER_MEMORY_HOP_1;
    int samples_to_process_2 = output_frames - FILTER_MEMORY_HOP_2;
    int start_point = 0;
    int i = 0;
    int j = 0;
    int k = 0;

    /* polyphase filter from 48kHz to 16kHz */

    /* first process the filter memory */
    start_point = 0;

    for (i = 0; i < FILTER_MEMORY_HOP_1; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < FILTER_MEMORY_1; j++, k++)
      {
	result += (int)(s->filter_memory_1[j]) * filter_coeffs_1[k];
      }

      for (j = 0; j < start_point + 1; j++, k++)
      {
	result += (int)(input[j]) * filter_coeffs_1[k];
      }

#ifdef USE_SATURATION
      temp_buffer[i] = (short)EAP_Clip16((result + 16384) >> 15);
#else
      temp_buffer[i] = (short)((result + 16384) >> 15);
#endif

      start_point += DOWNSAMPLING_FACTOR_1;
    }

    output_samples = &temp_buffer[FILTER_MEMORY_HOP_1];
    poly_start = &input[0];

    /* then process the rest of the input buffer */
    for (i = 0; i < samples_to_process_1; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_1; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_1[j];
      }

#ifdef USE_SATURATION
      *output_samples++ = (short)EAP_Clip16((result + 16384) >> 15);
#else
      *output_samples++ = (short)((result + 16384) >> 15);
#endif

      poly_start += DOWNSAMPLING_FACTOR_1;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->filter_memory_1[0])),
	   (void*)poly_start,
	   FILTER_MEMORY_1 * sizeof(short));

    /* polyphase filter from 16kHz to 8kHz */

    /* first process the filter memory */
    start_point = 0;
    for (i = 0; i < FILTER_MEMORY_HOP_2; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < FILTER_MEMORY_2; j++, k++)
      {
	result += (int)(s->filter_memory_2[j]) * filter_coeffs_2[k];
      }

      for (j = 0; j < start_point + 1; j++, k++)
      {
	result += (int)(temp_buffer[j]) * filter_coeffs_2[k];
      }

#ifdef USE_SATURATION
      output[i] = (short)EAP_Clip16((result + 16384) >> 15);
#else
      output[i] = (short)((result + 16384) >> 15);
#endif

      start_point += DOWNSAMPLING_FACTOR_2;
    }

    output_samples = &output[FILTER_MEMORY_HOP_2];
    poly_start = &temp_buffer[1];

    /* then process the rest of the temp buffer */
    for (i = 0; i < samples_to_process_2; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_2; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_2[j];
      }

#ifdef USE_SATURATION
      *output_samples++ = (short)EAP_Clip16((result + 16384) >> 15);
#else
      *output_samples++ = (short)((result + 16384) >> 15);
#endif

      poly_start += DOWNSAMPLING_FACTOR_2;
    }

    /* copy filter memory */
    memcpy((void*)(&(s->filter_memory_2[0])),
           (void*)poly_start,
           FILTER_MEMORY_2 * sizeof(short));

    return output_frames;
}

int process_src_48_to_8_stereo_to_mono(src_48_to_8 *s,
				       short *output,
				       short *input,
				       int input_frames)
{
    signed short *input_samples = 0;
    signed short *poly_start = 0;
    signed short *output_samples = 0;
    signed int result = 0;
    int intermediate_frames = input_frames / 6;
    int output_frames = input_frames / 12;
    int samples_to_process_1 = intermediate_frames - FILTER_MEMORY_HOP_1;
    int samples_to_process_2 = output_frames - FILTER_MEMORY_HOP_2;
    int start_point = 0;
    int i = 0;
    int j = 0;
    int k = 0;

    /* polyphase filter from 48kHz to 16kHz */

    /* first process the filter memory */
    start_point = 0;

    for (i = 0; i < FILTER_MEMORY_HOP_1; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < FILTER_MEMORY_1 * 2; j += 2, k++)
      {
	result += (int)(s->filter_memory_1[j]) * filter_coeffs_1[k];
      }

      for (j = 0; j < (start_point + 1) * 2; j += 2, k++)
      {
	result += (int)(input[j]) * filter_coeffs_1[k];
      }

#ifdef USE_SATURATION
      temp_buffer[i] = (short)EAP_Clip16((result + 16384) >> 15);
#else
      temp_buffer[i] = (short)((result + 16384) >> 15);
#endif

      start_point += STEREO_DOWNSAMPLING_FACTOR_1;
    }

    output_samples = &temp_buffer[FILTER_MEMORY_HOP_1];
    poly_start = &input[0];

    /* then process the rest of the input buffer */
    for (i = 0; i < samples_to_process_1; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_1; j++)
      {
	result += (int)(*input_samples) * filter_coeffs_1[j];
	input_samples += 2;
      }

#ifdef USE_SATURATION
      *output_samples++ = (short)EAP_Clip16((result + 16384) >> 15);
#else
      *output_samples++ = (short)((result + 16384) >> 15);
#endif

      poly_start += STEREO_DOWNSAMPLING_FACTOR_1;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->filter_memory_1[0])),
	   (void*)poly_start,
	   2 * FILTER_MEMORY_1 * sizeof(short));

    /* polyphase filter from 16kHz to 8kHz */

    /* first process the filter memory */
    start_point = 0;
    for (i = 0; i < FILTER_MEMORY_HOP_2; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < FILTER_MEMORY_2; j++, k++)
      {
	result += (int)(s->filter_memory_2[j]) * filter_coeffs_2[k];
      }

      for (j = 0; j < start_point + 1; j++, k++)
      {
	result += (int)(temp_buffer[j]) * filter_coeffs_2[k];
      }

#ifdef USE_SATURATION
      output[i] = (short)EAP_Clip16((result + 16384) >> 15);
#else
      output[i] = (short)((result + 16384) >> 15);
#endif

      start_point += DOWNSAMPLING_FACTOR_2;
    }

    output_samples = &output[FILTER_MEMORY_HOP_2];
    poly_start = &temp_buffer[1];

    /* then process the rest of the temp buffer */
    for (i = 0; i < samples_to_process_2; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_2; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_2[j];
      }

#ifdef USE_SATURATION
      *output_samples++ = (short)EAP_Clip16((result + 16384) >> 15);
#else
      *output_samples++ = (short)((result + 16384) >> 15);
#endif

      poly_start += DOWNSAMPLING_FACTOR_2;
    }

    /* copy filter memory */
    memcpy((void*)(&(s->filter_memory_2[0])),
           (void*)poly_start,
           FILTER_MEMORY_2 * sizeof(short));

    return output_frames;
}

#if 0
int process_src_48_to_8_stereo_to_mono(src_48_to_8 *s, short *output, short *input, int input_frames)
{
    signed short *input_samples = 0;
    signed short *poly_start = 0;
    signed short *output_samples = 0;
    signed int result = 0;
    int intermediate_frames = input_frames / 6;
    int output_frames = input_frames / 12;
    int start_point = 0;
    int i = 0;
    int j = 0;
    int k = 0;

    /* polyphase filter from 48kHz to 16kHz */

    /* first process the filter memory */
    start_point = 0;
    for (i = 0; i < 5; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < (FILTER_LENGTH_1 - 1) * 2; j += 2, k++)
      {
	result += (int)(s->filter_memory_1[j]) * filter_coeffs_1[k];
      }

      for (j = 0; j < (start_point + 1) * 2; j += 2, k++)
      {
	result += (int)(input[j]) * filter_coeffs_1[k];
      }

      temp_buffer[i] = (short)((result + 16384) >> 15);

      start_point += DOWNSAMPLING_FACTOR_1 * 2;
    }

    output_samples = &temp_buffer[5];
    poly_start = &input[0];

    /* then process the rest of the input buffer */
    for (i = 0; i < intermediate_frames - 5; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_1; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_1[j];
	input_samples++;
      }

      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_1 * 2;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->filter_memory_1[0])),
	   (void*)(&(input[465 * 2])),
	   2 * 15 * sizeof(short));

    /* polyphase filter from 16kHz to 8kHz */

    /* first process the filter memory */
    start_point = 0;
    for (i = 0; i < 40; i++)
    {
      result = 0;

      for (j = start_point, k = 0; j < FILTER_LENGTH_2 - 1; j++, k++)
      {
	result += (int)(s->filter_memory_2[j]) * filter_coeffs_2[k];
      }

      for (j = 0; j < start_point + 1; j++, k++)
      {
	result += (int)(temp_buffer[j]) * filter_coeffs_2[k];
      }

      output[i] = (short)((result + 16384) >> 15);
      start_point += DOWNSAMPLING_FACTOR_2;
    }

    output_samples = &output[40];
    poly_start = &temp_buffer[1];

    /* then process the rest of the temp buffer */
    for (i = 0; i < 40; i++)
    {
      result = 0;
      input_samples = poly_start;

      for (j = 0; j < FILTER_LENGTH_2; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_2[j];
      }

      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_2;
    }

    /* copy filter memory */
    memcpy((void*)(&(s->filter_memory_2[0])),
           (void*)(&(temp_buffer[81])),
           79 * sizeof(short));

    return output_frames;
}

int process_src_48_to_8(src_48_to_8 *s, short *output, short *input, int input_frames)
{
    signed short *input_samples = 0;
    signed short *poly_start = 0;
    signed short *output_samples = 0;
    signed int result = 0;
    int intermediate_frames = (input_frames/3);
    int output_frames = (input_frames/6);
    int i = 0;
    int j = 0;

    /* copy input to scratch buffer to make looping easier */
    memcpy((void*)(&(s->input_temp_buffer_1[15])),
	   input,
	   input_frames * sizeof(short));

    input_samples = &s->input_temp_buffer_1[0];
    output_samples = &s->input_temp_buffer_2[79];
    poly_start = &s->input_temp_buffer_1[0];

    /* polyphase filter from 48kHz to 16kHz */
    for (i = 0; i < intermediate_frames; i++)
    {
      result = 0;
      input_samples = poly_start;
      for (j = 0; j < FILTER_LENGTH_1; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_1[j];
      }
      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_1;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->input_temp_buffer_1[0])),
	   (void*)(&(s->input_temp_buffer_1[input_frames])),
	   15 * sizeof(short));

    input_samples = &s->input_temp_buffer_2[0];
    output_samples = output;
    poly_start = &s->input_temp_buffer_2[0];

    /* polyphase filter from 16kHz to 8kHz */
    for (i = 0; i < output_frames; i++)
    {
      result = 0;
      input_samples = poly_start;
      for (j = 0; j < FILTER_LENGTH_2; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_2[j];
      }
      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_2;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->input_temp_buffer_2[0])),
           (void*)(&(s->input_temp_buffer_2[intermediate_frames])),
           79 * sizeof(short));

    return output_frames;
}


int process_src_48_to_8_stereo_to_mono(src_48_to_8 *s, short *output, short *input, int input_frames)
{
    signed short *input_samples = 0;
    signed short *poly_start = 0;
    signed short *output_samples = 0;
    signed int result = 0;
    int intermediate_frames = (input_frames/3);
    int output_frames = (input_frames/6);
    int i = 0;
    int j = 0;

    /* copy only the interleaved 2nd channel */
    for (i = 0; i < input_frames; i++)
    {
      s->input_temp_buffer_1[15+i] = input[i*2];
    }

    input_samples = &s->input_temp_buffer_1[0];
    output_samples = &s->input_temp_buffer_2[79];
    poly_start = &s->input_temp_buffer_1[0];

    /* polyphase filter from 48kHz to 16kHz */
    for (i = 0; i < intermediate_frames; i++)
    {
      result = 0;
      input_samples = poly_start;
      for (j = 0; j < FILTER_LENGTH_1; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_1[j];
      }
      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_1;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->input_temp_buffer_1[0])),
	   (void*)(&(s->input_temp_buffer_1[input_frames])),
	   15 * sizeof(short));

    input_samples = &s->input_temp_buffer_2[0];
    output_samples = output;
    poly_start = &s->input_temp_buffer_2[0];

    /* polyphase filter from 16kHz to 8kHz */
    for (i = 0; i < output_frames; i++)
    {
      result = 0;
      input_samples = poly_start;
      for (j = 0; j < FILTER_LENGTH_2; j++)
      {
	result += (int)(*input_samples++) * filter_coeffs_2[j];
      }
      *output_samples++ = (short)((result + 16384) >> 15);
      poly_start += DOWNSAMPLING_FACTOR_2;
    }

    /* copy filter memory to the beginning of scratch buffer */
    memcpy((void*)(&(s->input_temp_buffer_2[0])),
           (void*)(&(s->input_temp_buffer_2[intermediate_frames])),
           79 * sizeof(short));

    return output_frames;
}
#endif
